src/channel.c

```diff
@@ -0,0 +1,232 @@
+#include "TH.h"
+#include "luaT.h"
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+#include <pthread.h>
+#include <errno.h>
+#include "ringbuffer.h"
+#include "serialize.h"
+#include "error.h"
+
+#define DEFAULT_CHANNEL_SIZE (16*1024)
+
+#define TOO_TRICKY (0)
+#define CHANNEL_VERBOSE (0)
+
+typedef struct channel_t {
+   struct ringbuffer_t* rb;
+   pthread_mutex_t mutex;
+   pthread_cond_t read_avail_cond;
+#if TOO_TRICKY
+   pthread_cond_t write_avail_cond;
+#endif
+   int closed;
+   int drained;
+   uint32_t num_items;
+   int refcount;
+   size_t size_increment;
+} channel_t;
+
+static void channel_init_queue(channel_t *channel, size_t size) {
+   // init queue mutex
+   pthread_mutexattr_t mutex_attr;
+   pthread_mutexattr_init(&mutex_attr);
+   pthread_mutexattr_settype(&mutex_attr, PTHREAD_MUTEX_RECURSIVE);
+   pthread_mutex_init(&channel->mutex, &mutex_attr);
+
+   // init condition variables
+   pthread_cond_init(&channel->read_avail_cond, NULL);
+#if TOO_TRICKY
```

Comment:
Do you want to actually keep this? I'm not sure Zak ever tested this extensively, no reason to propagate it to another file.